### PR TITLE
Driver: add support for the WebAssembly toolchain

### PR DESCRIPTION
**Pull request by MaxDesiatov**
`WebAssemblyToolChains.cpp` is added to the Driver with a basic toolchain for WebAssembly. This supports building and linking Swift code that targets WebAssembly.

Resolves a part of SR-9307.

CC @kateinoigakukun

### CODE REVIEW

**Comment by MaxDesiatov** [comment]
@swift-ci please smoke test

File: lib/Driver/WebAssemblyToolChains.cpp [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+#include "llvm/Support/Path.h"
+#include "llvm/Support/Process.h"
+#include "llvm/Support/Program.h"
+
+using namespace swift;
+using namespace swift::driver;
+using namespace llvm::opt;
+
+std::string toolchains::WebAssembly::sanitizerRuntimeLibName(StringRef Sanitizer,
+                                                         bool shared) const {
```
Comment (kateinoigakukun) [inline lib/Driver/WebAssemblyToolChains.cpp:42]: Could you apply clang-format?

File: lib/Driver/WebAssemblyToolChains.cpp [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+  assert(context.Output.getPrimaryOutputType() == file_types::TY_Image &&
+         "Invalid linker output type.");
+
+  ArgStringList Arguments;
+
+  std::string Target = getTriple().str();
+  if (!Target.empty()) {
+    Arguments.push_back("-target");
+    Arguments.push_back(context.Args.MakeArgString(Target));
+  }
```
Comment (kateinoigakukun) [inline lib/Driver/WebAssemblyToolChains.cpp:79]: `--target` option is added twice https://github.com/apple/swift/pull/39299/files#diff-678e83f2487547ddadcb72231e5d6a8dd296c0fb001889af4aba5271f23420abR161-R163

File: lib/Driver/WebAssemblyToolChains.cpp [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+
+  // WebAssembly doesn't reserve low addresses But without "extra inhabitants"
+  // of the pointer representation, runtime performance and memory footprint are
+  // worse. So assume that compiler driver uses wasm-ld and --global-base=1024
+  // to reserve low 1KB.
+  Arguments.push_back("-Xlinker");
+  Arguments.push_back(context.Args.MakeArgString(
+      Twine("--global-base=") +
+      std::to_string(SWIFT_ABI_WASM32_LEAST_VALID_POINTER)));
```
Comment (kateinoigakukun) [inline lib/Driver/WebAssemblyToolChains.cpp:197]: I think this macro is not defined in upstream yet.

**Review by kateinoigakukun — CHANGES_REQUESTED** [review]

File: lib/Driver/WebAssemblyToolChains.cpp (hunk 3 above)
Comment (MaxDesiatov) [reply lib/Driver/WebAssemblyToolChains.cpp]: I removed it for now from this driver code. We can add it in a separate follow-up PR.

**Comment by MaxDesiatov** [comment]
@swift-ci please smoke test

File: lib/Driver/WebAssemblyToolChains.cpp (hunk 2 above)
Comment (MaxDesiatov) [reply lib/Driver/WebAssemblyToolChains.cpp:81]: Good catch, I cleaned that up.

**Comment by MaxDesiatov** [comment]
@swift-ci please smoke test

**Comment by kateinoigakukun** [comment]
Could you add driver test for this toolchain using `--driver-print-jobs` ?

**Comment by MaxDesiatov** [comment]
@swift-ci please smoke test

**Comment by MaxDesiatov** [comment]
@swift-ci please smoke test

**Comment by MaxDesiatov** [comment]
@swift-ci please smoke test

**Comment by MaxDesiatov** [comment]
@swift-ci please smoke test

**Comment by MaxDesiatov** [comment]
@swift-ci please smoke test

**Comment by MaxDesiatov** [comment]
@swift-ci please smoke test

**Review by kateinoigakukun — APPROVED** [review]

**Review by artemcm — APPROVED** [review]
The changes here look good to me. Sorry for the delay. 
And thank you @MaxDesiatov!

**Comment by artemcm** [comment]
@swift-ci please test

**Pull request merged**